Collect file extension stats in `gh-pages`

abitrolly opened the pull request.

abitrolly committed:
Fix extensions statistics script

Skips filenames without extensions and strips dirs from them

abitrolly committed:
Update README.md with more details and link to raw data

abitrolly committed:
Compare git file stats over time

abitrolly committed:
Add manual trigger for debugging actions

abitrolly committed:
Speed up git when Firefox checkout is present

abitrolly committed:
Add git file extension stats to data

abitrolly committed:
Sort before counting as `uniq -c` counts only continuous ranges

abitrolly committed:
Extension stats are now in `gh-pages` branch

abitrolly committed:
Collect only single extensions

Because there is too much noise with double and more extensions

abitrolly commented:
Counting lines of code takes 3 seconds and calculating file extensions takes 3 minutes.

![image](https://user-images.githubusercontent.com/8781107/170021846-0599490f-5f9e-4c41-b9df-75d4e5161042.png)

It appears that `git ls-files | xargs -n 1 basename > 2` command is very slow - 3m14s compared to `git ls-files > 1` which takes only 2.2s.

abitrolly commented:
Firefox repo contains almost 22 million files, so that means `xargs` needs to run external process 22 million times. Looks like a major bottleneck for all shell pipelines.
```
✗ git ls-files | wc -c
21857117
```

abitrolly committed:
`xargs` is slow to invoke 20m subprocesses

Replacing with a single `grep` call.

 -P  - use Perl regexp
 -o  - output only matched
 (?<=) - look behind (don't include this part)

abitrolly committed:
Rename target file to extensions.txt

abitrolly committed:
Ident JSON

abitrolly committed:
Use CI independent script gather repo metrics

abitrolly committed:
Use absolute path for a build dir

abitrolly committed:
Rename workflow to distinguish in web UI

abitrolly committed:
Get historical `gecko-dev` commits

4e6 commented on .github/workflows/deploy.yml:
```diff
@@ -1,10 +1,11 @@
     branches:
       - main
   pull_request:
+  workflow_dispatch:
```
Why do we need it?

abitrolly replied:
For testing workflow in branches other than `main`.

4e6 commented on .github/workflows/deploy.yml:
```diff
@@ -29,6 +30,10 @@ jobs:
+          
+      - name: Gather commits by day and file extension statistics
+        run: |
+          ./01stats.sh gecko-dev build
```
What do we get out of this build step? Also I see that it takes pretty long time to complete.

abitrolly replied:
The goal is to move stats collection commands out of GitHub Actions YAML, so that they could be run standalone.

As I continued experiments in my `main` branch after opening this PR, more things started to creep in. The command that takes the most time is `git fetch --unshallow` added in the last commit to start playing with historical data.

4e6 commented:
Hi! Thanks for your contribution. Just curious, what's the motivation behind these changes?

abitrolly commented:
@4e6 well, this PR is far from being finished. The final goal was to get the dataset for diagrams on Firefox Oxidization over time. (like in #10). Because I didn't know the codebase, I started with the code that seemed to be the easiest to get up. Like counting file extensions over time.

Because only full git checkout takes the whole 15 minutes, going commit over commit probably won't be feasible to do in one CI run, so the plan is to collect the data month by month over multiple CI runs.

abitrolly committed:
Clone complete git history with latest action version

Maybe it will be faster that restoring `gecko-dev` history with,
`git fetch --unshallow`, which takes about 15 minutes.

abitrolly commented:
Doing complete `gecko-dev` checkout through the action, and it tool 10 minutes, where the stats script took only 12s. That's still too slow. Maybe the initial checkout can be cached.

![image](https://user-images.githubusercontent.com/8781107/171702639-5ff99839-1fd1-4960-b9c3-971412afa971.png)

https://github.com/abitrolly/firefox-lang-stats/runs/6713143940

abitrolly commented:
Opened the issue in https://github.com/actions/checkout/issues/818 to maybe track possible solutions.

abitrolly committed:
Try to explicitly checkout only master branch

https://github.com/actions/checkout#usage

abitrolly committed:
Try to clone Firefox without actions/checkout

Clone by action with specified branch was 1m slower
https://github.com/actions/checkout/issues/818#issuecomment-1145535053
